runtime/altair/src/migrations.rs

```diff
@@ -166,3 +169,145 @@ mod asset_registry {
 		}
 	}
 }
+
+mod orml_tokens_migration {
+	use cfg_primitives::AccountId;
+	use orml_tokens::AccountData;
+
+	use super::*;
+
+	/// As we dropped `CurrencyId::KSM` and `CurrencyId::AUSD`, we need to
+	/// migrate the balances under the dropped variants in favour of the new,
+	/// corresponding `CurrencyId::ForeignAsset`. We have never transferred KSM
+	/// so we only need to deal with AUSD.
+	pub struct CurrencyIdRefactorMigration;
+
+	#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode)]
+	pub struct OldState {
+		pub total_issuance: Balance,
+		pub entries: Vec<(AccountId, AccountData<Balance>)>,
+	}
+
+	impl OnRuntimeUpgrade for CurrencyIdRefactorMigration {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+			let total_issuance =
+				orml_tokens::TotalIssuance::<Runtime>::get(DEPRECATED_AUSD_CURRENCY_ID);
+			let entries: Vec<(AccountId, AccountData<Balance>)> =
+				orml_tokens::Accounts::<Runtime>::iter()
+					.filter(|(_, old_currency_id, _)| {
+						*old_currency_id == DEPRECATED_AUSD_CURRENCY_ID
+					})
+					.map(|(account, _, account_data)| (account, account_data))
+					.collect::<_>();
+
+			Ok(OldState {
+				total_issuance,
+				entries,
+			}
+			.encode())
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(state: Vec<u8>) -> Result<(), &'static str> {
+			use crate::OrmlTokens;
+
+			let old_state = OldState::decode(&mut state.as_ref())
+				.map_err(|_| "Error decoding pre-upgrade state")?;
+
+			let new_total_issuance =
+				orml_tokens::TotalIssuance::<Runtime>::get(NEW_AUSD_CURRENCY_ID);
+
+			ensure!(
+				old_state.total_issuance == new_total_issuance,
+				"The old AUSD issuance differs from the new one"
+			);
+
+			for (account, account_data) in old_state.entries {
+				ensure!(
+					OrmlTokens::accounts(&account, NEW_AUSD_CURRENCY_ID) == account_data.clone(),
+					"The account data under the new AUSD Currency does NOT match the old one"
+				);
+			}
+
+			return Ok(());
```

Comment:
Nit: The dark side (Solidity) has shown its impact!

```suggestion
			Ok(())
```

Reply:
Ai 😱  thanks!